### ISSUE TITLE
Decreasing write capacity was disallowed without any reasons, decreasing using Management Console worked

### ISSUE THREAD

**Issue by chris4826** [channl/dynamodb-lambda-autoscale#56]
Hi,

at first thank's for your great project. I have already tried it a lot using different scaling policies.
Today the write capacity was not decreased although 0 of 100 provisioned write capacity units have been used. This is the CloudWatch log:
...is consuming 0 of 100 (0%) write capacity units and is below minimum threshold of 20% so a decrement is WANTED but is DISALLOWED
At exactly this time the read capacity has been decreased. Thus there have been enough decreases left for this calendar day. This is really confusing, so I started to debug.
I ended in the file src/Provisioner.js containing this line:

`let isAdjustmentAllowed = isAfterLastDecreaseGracePeriod && isAfterLastIncreaseGracePeriod && isReadDecrementAllowed;`

Why is only ReadDecrement considered? Why is there no "isWriteDecrementAllowed" using the function calculateDecrementedWriteCapacityValue?
Maybe it's bug and this explains why decreasing read capacity has been allowed whereas decreasing write capacity was denied.

You find my config DefaultProvisioner.js as an attachment.

Thank's for your help.

Best regards,
Chris
[DefaultProvisioner.txt](https://github.com/channl/dynamodb-lambda-autoscale/files/1039141/DefaultProvisioner.txt)

**Comment by tmitchel2** [comment]
Hey, yes you're completely right, it is a bug.

I actually noticed that recently and have fixed it on an experimental branch where i am performing a large refactor.  https://github.com/channl/dynamodb-lambda-autoscale/blob/experimental/src/ProvisionedThroughtputCalculator.js You can see my change here.  Unfortunately just at the moment I have zero spare time to look into this but happy to accept a simple pull request.  Otherwise i will soon be merging my refactor to main once fully tested.

**Comment by chris4826** [comment]
Hi,
thank's for your feedback.
I compared the files and tried to find a solution.
Unfortunately I have to use your commit dccd22440b33cc3f0167bd1ca876a7f7d4a10600 because the newer commits do not work for me. You have reduced the number of packages in packages.json. When running "npm run build" I'm getting a lot of error, e.g. "gulp dist" not found. Do you already know these problems, too?

So I tried to fix the issue with the old commit where src/utils/ThroughputUtils.js doesn't exists. But src/utils/Throughput.js consists of similar code.

I introduced a new variable for writes:
`    let isWriteDecrementAllowed = adjustmentContext.AdjustmentType === 'decrement' ?
      RateLimitedDecrement.isDecrementAllowed(data, adjustmentContext, d => this.calculateDecrementedWriteCapacityValue(d)) :
      true;
`

And then changed isAdjustmentAllowed with an addtional or operator:
`
    let isAdjustmentAllowed = isAfterLastDecreaseGracePeriod && isAfterLastIncreaseGracePeriod && (isReadDecrementAllowed || isWriteDecrementAllowed);
`


I will let you know if this small fix works properly after doing some load test. If you already know that this won't work please let me know.

**Comment by tmitchel2** [comment]
Yea that doesn't look quite right to me because you would allow a decrement on read if either read or write decrement is allowed, and then the same for writes.

**Comment by chris4826** [comment]
I agree with you. I will probably wait till you merge the experimental branch into master because your refactoring is too big in order to select only specific piece of code from the experimental branch.
